### ISSUE TITLE
[Bug]: 7.1 Stories are leaking headers into table of contents

### ISSUE THREAD

**Issue by ajkl2533** [storybookjs/storybook#23541]
### Describe the bug

When I enable the table of contents the headers from within stories are displayed in the TOC. I would expect that stories are isolated from the rest of the storybook.

<img width="1260" alt="Screenshot 2023-07-20 at 16 25 36" src="https://github.com/storybookjs/storybook/assets/18654425/3af987fa-1617-4f52-9cdb-6269333a419e">

### To Reproduce

1. Enable TOC
2. Create a story and use `h3` tag (depends on configuration) in the story

### System

```shell
Environment Info:

  System:
    OS: macOS 13.4.1
    CPU: (10) arm64 Apple M1 Pro
  Binaries:
    Node: 16.16.0 - ~/.nvm/versions/node/v16.16.0/bin/node
    Yarn: 3.5.0 - ~/.nvm/versions/node/v16.16.0/bin/yarn
    npm: 8.14.0 - ~/Projects/design-system-v2/node_modules/.bin/npm
  Browsers:
    Chrome: 114.0.5735.198
    Safari: 16.5.1
  npmPackages:
    @storybook/addon-a11y: ^7.1.0 => 7.1.0 
    @storybook/addon-actions: ^7.1.0 => 7.1.0 
    @storybook/addon-designs: ^7.0.1 => 7.0.1 
    @storybook/addon-docs: ^7.1.0 => 7.1.0 
    @storybook/addon-essentials: ^7.1.0 => 7.1.0 
    @storybook/addon-interactions: ^7.1.0 => 7.1.0 
    @storybook/blocks: ^7.1.0 => 7.1.0 
    @storybook/components: ^7.1.0 => 7.1.0 
    @storybook/jest: ^0.1.0 => 0.1.0 
    @storybook/react: ^7.1.0 => 7.1.0 
    @storybook/react-vite: ^7.1.0 => 7.1.0 
    @storybook/test-runner: ^0.11.0 => 0.11.0 
    @storybook/testing-library: ^0.2.0 => 0.2.0
```

**Comment by vmizgiris-equinix** [comment]
@ajkl2533, I was able to get around this by explicitly excluding headers from Story blocks in the TOC config:

```js
docs: {
  toc: {
    disable: false,
    headingSelector: 'h2, h3',
    ignoreSelector: '.docs-story h2, .docs-story h3',
  },
}
```

You just have to match the heading tags in the `headingSelector` with the `ignoreSelector`. This should still probably be fixed at the library level.

**Comment by shilman** [comment]
@vmizgiris-equinix Any chance you can contribute a PR with the fix?

**Comment by vmizg** [comment]
> @vmizgiris-equinix Any chance you can contribute a PR with the fix?

PR raised: #23677. Let me know if the suggested approach is okay.
I was initially tinkering with prefixing the docs-story class name to all given heading selectors, but it seems to be cleaner and good for a default to just do `.docs-story *`.

**Comment by BAHAA-THE-KING** [comment]
Hello
I am using v7.6.16 and the issue still present, here is an image :
![Screenshot_6](https://github.com/storybookjs/storybook/assets/98487604/e5625c43-0056-4da1-850f-738c609de8d1)

Note that this:
`ignoreSelector: '.docs-story h2, .docs-story h3',`
solve it, but can't it be done automatically ?

**Comment by vmizg** [comment]
@BAHAA-THE-KING the fix is in 7.6.17
